### PR TITLE
ByteString arrayOffset method

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The ByteString class currently assumes the underlying array will be a complete representation of data. This is limiting as it does not allow a subsection of another array to be used. The forces copy operations to take place to compensate for the lack of API support.

Modifications:
- add arrayOffset method to ByteString
- modify all ByteString and AsciiString methods that loop over or index into the underlying array to use this offset
- update all code that uses ByteString.array to ensure it accounts for the offset
- add unit tests to test the implementation respects the offset

Result:
ByteString and AsciiString can represent a sub region of a byte[].

### CODE REVIEW

**Comment by Scottmitch** [comment]
@trustin @nmittler @normanmaurer @buchgr @rkapsi - PTAL.

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 1]
```diff
@@ -541,24 +540,32 @@ public void copy(int srcIdx, char[] dst, int dstIdx, int length) {
...
         final int dstEnd = dstIdx + length;
-        for (int i = srcIdx, j = dstIdx; j < dstEnd; i++, j++) {
-            dst[j] = b2c(value[i]);
+        for (int i = dstIdx, j = srcIdx + arrayOffset(); i < dstEnd; i++, j++) {
+            dst[i] = b2c(value[j]);
         }
     }
 
     @Override
+    public AsciiString subSequence(int start) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: Are these casts safe ?

File: common/src/main/script/codegen.groovy [hunk 2]
```diff
@@ -23,7 +23,7 @@ void convertTemplates(String templateDir,
     def keyName = keyPrimitive.capitalize()
     def replaceFrom = "(^.*)K([^.]+)\\.template\$"
     def replaceTo = "\\1" + keyName + "\\2.java"
-    def hashCodeFn = keyPrimitive.equals("long") ? "(int)(key ^ (key >>> 32))" : "(int) key"
+    def hashCodeFn = keyPrimitive.equals("long") ? "(int) (key ^ (key >>> 32))" : "(int) key"
```
Comment (normanmaurer) [inline common/src/main/script/codegen.groovy]: not really related to this change...

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 1 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: Humm...I think so. If this method is called the type must be of type `AsciiString`, and the underlying `subSequence(..)` implementation should be fine because it is just working on the byte array bcz of the reduced charset.  Am I missing something?

File: common/src/main/script/codegen.groovy (hunk 2 above)
Comment (Scottmitch) [inline common/src/main/script/codegen.groovy]: Agreed...I copy/pasted this code for the `PlatformDependent0.hashCode(...)` and got a style compile error.  I guess we wouldn't catch that with this code because it generating code.  That being said I could pull this out...or just leave it and it will get fixed when I merge forward.

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 3]
```diff
@@ -349,6 +349,28 @@ static boolean equals(byte[] bytes1, int startPos1, int endPos1, byte[] bytes2,
         return bytes1[startPos1] == bytes2[startPos2];
     }
 
+    static int hashCode(byte[] bytes, int startPos, int endPos, long prime) {
```
Comment (buchgr) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @Scottmitch can you add a test that shows that this function and `safeHashCode` are equivalent? (as good as this is possible with a test).

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 4]
```diff
@@ -69,11 +68,10 @@ public int compare(CharSequence o1, CharSequence o2) {
             int length2 = o2.length();
             int minLength = Math.min(length1, length2);
             if (a1 != null && a2 != null) {
-                byte[] thisValue = a1.value;
-                byte[] thatValue = a2.value;
-                for (int i = 0; i < minLength; i++) {
-                    byte v1 = thisValue[i];
-                    byte v2 = thatValue[i];
+                final int a1Len = minLength + a1.arrayOffset();
```
Comment (buchgr) [inline common/src/main/java/io/netty/util/AsciiString.java]: @Scottmitch can you add a benchmark that shows that keeping this large amounts of code is significantly faster than just using a single loop with `charAt`?

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 1 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: Maybe I'm missing something but the super.subSequence below will create a new ByteString which will cause then a ClassCastException if I don't have a brain fart:

https://github.com/netty/netty/blob/master/common/src/main/java/io/netty/util/ByteString.java#L420

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: oof...you are right.  I will address this.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: Thanks! I almost thought my brain is already in weekend-mode ;) Please also add a unit test for this

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java (hunk 3 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @buchgr +1.

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 4 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: Working on it...I'll cpy/paste the code you have in your PR to compare against.

Comment (buchgr) [inline common/src/main/java/io/netty/util/AsciiString.java]: @Scottmitch thanks!

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java (hunk 3 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: Great suggestion BTW :)  Rushed to get through approval process and "safe" method was not so safe in this sense...take a look at the code now.

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 1 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: Yip done. take a look

**Comment by Scottmitch** [comment]
@buchgr - Here you go:

```Benchmark                                       (size)   Mode  Cnt         Score        Error  Units
AsciiStringBenchmark.complexInsensativeCompare      10  thrpt   20  25881070.811 ± 143484.632  ops/s
AsciiStringBenchmark.complexInsensativeCompare     100  thrpt   20   3898435.486 ±  17824.358  ops/s
AsciiStringBenchmark.complexInsensativeCompare    1000  thrpt   20    485045.614 ±    367.934  ops/s
AsciiStringBenchmark.complexInsensativeCompare   10000  thrpt   20     48434.538 ±    574.461  ops/s
AsciiStringBenchmark.complexInsensativeCompare  100000  thrpt   20      4917.089 ±      5.499  ops/s
AsciiStringBenchmark.complexSensativeCompare        10  thrpt   20  39311729.069 ± 215287.538  ops/s
AsciiStringBenchmark.complexSensativeCompare       100  thrpt   20  11358090.386 ±  16428.000  ops/s
AsciiStringBenchmark.complexSensativeCompare      1000  thrpt   20   1305740.981 ±   2166.293  ops/s
AsciiStringBenchmark.complexSensativeCompare     10000  thrpt   20    141399.231 ±    176.152  ops/s
AsciiStringBenchmark.complexSensativeCompare    100000  thrpt   20     14120.667 ±      8.289  ops/s
AsciiStringBenchmark.simpleInsensativeCompare       10  thrpt   20  21860398.628 ± 275208.059  ops/s
AsciiStringBenchmark.simpleInsensativeCompare      100  thrpt   20   3055570.415 ±  53073.958  ops/s
AsciiStringBenchmark.simpleInsensativeCompare     1000  thrpt   20    318989.668 ±   9318.473  ops/s
AsciiStringBenchmark.simpleInsensativeCompare    10000  thrpt   20     22072.309 ±    102.826  ops/s
AsciiStringBenchmark.simpleInsensativeCompare   100000  thrpt   20      1314.410 ±      4.863  ops/s
AsciiStringBenchmark.simpleSensativeCompare         10  thrpt   20  38683486.474 ± 176874.275  ops/s
AsciiStringBenchmark.simpleSensativeCompare        100  thrpt   20  10618091.726 ±  31343.587  ops/s
AsciiStringBenchmark.simpleSensativeCompare       1000  thrpt   20   1367593.426 ±   1637.948  ops/s
AsciiStringBenchmark.simpleSensativeCompare      10000  thrpt   20    143448.066 ±    210.082  ops/s
AsciiStringBenchmark.simpleSensativeCompare     100000  thrpt   20     14408.861 ±     12.843  ops/s
```

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java (hunk 3 above)
Comment (buchgr) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: @Scottmitch 

Oh, so now you changed the hash function :-). How can you proof that this new hash function produces a better hash code or at least isn't likely to have more collisions? It might evaluate faster, but it might also produce a worse hash code. That's why I was hesitant to change the hash function with my equals performance update.

Looking at `j.l.String` and Hotspot, they are calculating the hash code character by character and it's not intrinsified by HotSpot. I think that's also what we should do. The hashcode is being cached after all and as I said in another comment, I haven't seen `AsciiString.hashCode()` show up in any profiler while working on the headers.

**Comment by buchgr** [comment]
@Scottmitch 

thanks! So I think it's fair to say that sizes bigger than 100 bytes we don't care about for AsciiString. Even 100 bytes is a bit much, but that's okay. (We mostly care about strings / headers in the 5 - 50 byte range). So for 10 bytes the case sensitive compare is 2% slower, the case insensitive 16%. For 100 bytes the case sensitive is 7% slower and the case insensitive 22%. Honestly, I would still go with the simple compare. That's a microbench, not a real application.

So my suggestion (as a compromise) in the other PR was to have one version for `CharSequence` and one version for `AsciiString`. Having one for all the possible combinations of the two seems a bit much.

**Comment by normanmaurer** [comment]
@netkins build

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersEncoder.java [hunk 5]
```diff
@@ -109,7 +109,10 @@ public Configuration configuration() {
     }
 
     private void encodeHeader(ByteString key, ByteString value, OutputStream stream) throws IOException {
-        encoder.encodeHeader(stream, key.array(), value.array(), sensitivityDetector.isSensitive(key, value));
+        encoder.encodeHeader(stream,
+                key.isEntireArrayUsed() ? key.array() : new ByteString(key, true).array(),
+                value.isEntireArrayUsed() ? value.array() : new ByteString(value, true).array(),
```
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersEncoder.java]: @Scottmitch, are you going to update the signature of the `Encoder.encodeHeader(stream, keyArray, valueArray, sensitivity)` so we don't need to copy?

**Comment by buchgr** [comment]
@normanmaurer @trustin @Scottmitch Before Java 7 `j.l.String` used to have this feature, but they removed it, afaik one reason being that it can lead to memory leaks (small substring referencing part of a big string). Just an FYI :-).

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersEncoder.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersEncoder.java]: @trustin - I would like to do this, and we have an issue to track it here https://github.com/netty/netty/issues/3597.  I think others are also interested in seeing these types of changes too.

**Comment by Scottmitch** [comment]
@buchgr - This was collapsed so I'll pull it out https://github.com/netty/netty/pull/3649#discussion_r28633871:

> @Scottmitch

> Oh, so now you changed the hash function :-). How can you proof that this new hash function produces a better hash code or at least isn't likely to have more collisions? It might evaluate faster, but it might also produce a worse hash code. That's why I was hesitant to change the hash function with my equals performance update.

> Looking at j.l.String and Hotspot, they are calculating the hash code character by character and it's not intrinsified by HotSpot. I think that's also what we should do. The hashcode is being cached after all and as I said in another comment, I haven't seen AsciiString.hashCode() show up in any profiler while working on the headers.

> Also how did you test the big endian version?

**Comment by Scottmitch** [comment]
@buchgr 

> How can you proof that this new hash function produces a better hash code or at least isn't likely to have more collisions?

This is an excellent point. My initial focus was just to to take the existing `AsciiString`'s hash code algorithm's XOR and and operations and extend that into the long/int/char realms....no mathematical proof was done but I started empirically testing.  I introduced a simple example to be able to distribute a work load of random byte hashing and collision detection and the results looked promising, but at the end of the day I didn't have any strong theoretical backing and just ended up porting Murmur3.  This hash algorithm is more widely used and understood to have a favorable performance / collision ratio.  I did some benchmarking below and seems to be about a 2x boost over `Arrays.hashCode(..)`.

```
Benchmark                                       (size)   Mode  Cnt         Score        Error  Units
PlatformDependentBenchmark.arraysBytesHashCode     100  thrpt   20  12680337.891 ± 369222.871  ops/s
PlatformDependentBenchmark.arraysBytesHashCode  100000  thrpt   20     12331.054 ±    300.674  ops/s
PlatformDependentBenchmark.unsafeBytesHashCode     100  thrpt   20  21647401.177 ± 845628.253  ops/s
PlatformDependentBenchmark.unsafeBytesHashCode  100000  thrpt   20     23767.495 ±    819.359  ops/s
```

> I haven't seen AsciiString.hashCode() show up in any profiler while working on the headers.

Despite this it may be a good idea to unify the hash codes in Netty to use more of a standard approach.  For example [ByteBufUtil](https://github.com/netty/netty/blob/master/buffer/src/main/java/io/netty/buffer/ByteBufUtil.java#L152) implements a custom hash code algorithm, and I'm not sure of its origins.  We would need to do evaluation but perhaps we could adopt Murmur3.

> Also how did you test the big endian version?

Big endian has not yet been tested and needs to be done.  Any suggestions for setting up an emulation environment?

**Comment by Scottmitch** [comment]
@normanmaurer @buchgr @nmittler @trustin - PTAL.  I dumped some more code here related to the hash code algorithm.  This also includes a new example....I can pull stuff out of this PR and open a new one if necessary.

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 6]
```diff
@@ -179,6 +172,54 @@ public int compare(CharSequence o1, CharSequence o2) {
...
+                if (result != 0) {
+                    return result;
+                }
+            }
+            return length1 - length2;
+        }
+    };
+
+    /**
+     * Factory which uses the {@link #ByteString(byte[], int, int, boolean)} constructor.
```
Comment (nmittler) [inline common/src/main/java/io/netty/util/AsciiString.java]: ByteString->AsciiString?

File: common/src/main/java/io/netty/util/ByteString.java [hunk 7]
```diff
@@ -317,36 +318,42 @@ public final int forEachByteDesc(ByteProcessor visitor) throws Exception {
...
             }
         }
         return -1;
     }
 
     public final byte byteAt(int index) {
-        return value[index];
+        // We must do a range check here to enforce the access does not go outside our sub region of the array.
+        // We rely on the array access itself to pick up the array out of bounds conditions
+        if (index < 0 || index >= length) {
```
Comment (nmittler) [inline common/src/main/java/io/netty/util/ByteString.java]: Would `length - index > 0` be slightly cheaper here?

**Comment by nmittler** [comment]
@Scottmitch LGTM

**Comment by buchgr** [comment]
@Scottmitch 
Seems like you were busy this weekend! Looks awesome! :-) Will review later today. Just three things immediately popping in my head
 1. Why don't you just fork an agreed-to-be-correct-and-fast Java implementation of murmur?
 2. You could have just used something like https://code.google.com/p/smhasher/ to evaluate your "unsafe hash function", maybe it was fine ... 
 3. What I tried to say in my previous comment is that we probably don't have to worry about having the same hash function for little and big endian systems. We could just fall back to the old "byte-by-byte" hash function for big endian, that way we at least can test it and know it performs reasonably well.

**Comment by Scottmitch** [comment]
@buchgr - Good questions :)

1. [Murmur3_32HashFunctio](https://github.com/google/guava/blob/bcd8abb54113d4ebd3f245d18da8af99ba67098f/guava/src/com/google/common/hash/Murmur3_32HashFunction.java) from guava works with CharSequence and their [known output values test](https://github.com/google/guava/blob/fa95e381e665d8ee9639543b99ed38020c8de5ef/guava-tests/test/com/google/common/hash/Murmur3Hash32Test.java#L45) which I borrowed from produced different results than the [Murmur 3 C++ Reference Implementation](http://smhasher.googlecode.com/svn/trunk/MurmurHash3.cpp) when I built/ran it.  None of the existing implementations I looked at accounted for endianness.
2. Didn't know about this project....good to know.  Could have been fine...but why invent another hash algorithm when there are already ones that will do what we need.
3. So I still need to test but the endianness code however the code to account for endianness isn't too pervasive and what is there should be easy for JIT to optimize (assumption based on the fact that the if branches will always evaluate the same way).  That being said haven't tested it...I'm tracking down a test environment....

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 6 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: done.

File: common/src/main/java/io/netty/util/ByteString.java (hunk 7 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/ByteString.java]: Can you be more explicit (do you mean `if (index - length > 0) { /*ERROR*/} `)? I'm not sure this completely covers all cases.  I need to make sure the index is in the range `[0, length)`.  However it seems like there are always boundary conditions where the integer arithmetic may wrap and give incorrect results.

**Comment by buchgr** [comment]
> So I still need to test but the endianness code isn't too pervasive and what is there should be easy for JIT to optimize (assumption based on the fact that the if branches will always evaluate the same way). That being said haven't tested it...I'm tracking down a test environment....

I am more worried about the code complexity and code size. Personally, I would only add this hash function for unsafe and little endian and if one of them is not available, go with the byte by byte hash code.

**Comment by Scottmitch** [comment]
@buchgr @nmittler - I have pushed the code which helps give some background for our discussion tomorrow.  The challenge was to introduce a consistent hash algorithm for case sensitive (just the byte value itself) and also case insensitive (requires translation).  These use cases we have also requires support for different data types (CharSequence, and ByteString/AsciiString).  The code is not complete (documentation needed, unit tests needed, etc...) but as I said it is a good place to start discussions for how this overlaps with (if at all) the headers updates @buchgr is investigating.

Just an FYI I still need to investigate the previous implementation of `caseInsensitiveHashCode` for a few reasons... 1. It doesn't look like there is any effort to make the hash "insensitive"...it looks the same as the regular hashCode(). 2. The cached value is being used if the value is AsciiString...which is the case "sensitive" value...unless everything was being translated to lowercase I'm missing something.

**Comment by Scottmitch** [comment]
@trustin @normanmaurer @nmittler @buchgr - PTAL.  I extracted out the example from this PR and will submit that as another PR.

@buchgr - I think I would like to keep the comparators as they are in this PR.  After you header improvements I may take a crack at removing them completely...or at least moving them to a `Util` type class and not using them for Headers.

File: common/src/main/java/io/netty/util/ByteString.java (hunk 7 above)
Comment (nmittler) [inline common/src/main/java/io/netty/util/ByteString.java]: Agreed ... keep it as-is.

**Comment by buchgr** [comment]
@Scottmitch 

I still don't see why we need a big endian and "unsafe is not supported" version. This complicates everything and is effectively dead code. Falling back to the current hash function would work just fine. I even don't understand why we need a new hash function ...

**Comment by Scottmitch** [comment]
@buchgr - I'm going to move the hash code stuff to a new PR to discuss separately....seemed to have support from a few folks but easier to discuss in isolation.

Note that case insensitive hash code is broken though...have that fixed in here...and will just submit the hash code stuff I have as a fix for that also.

**Comment by Scottmitch** [comment]
@buchgr @nmittler - I pulled the hash code changes out of this PR.  The scope should now be limited to the original intention of this PR.  PTAL.

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 4 above)
Comment (nmittler) [inline common/src/main/java/io/netty/util/AsciiString.java]: @Scottmitch have you put together the benchmarks for this?

**Comment by nmittler** [comment]
@Scottmitch LGTM, just one question about a benchmark that @buchgr was asking for.

**Comment by Scottmitch** [comment]
@nmittler - Can you provide a link for the question?  I don't see it.

**Comment by nmittler** [comment]
Sure https://github.com/netty/netty/pull/3649/files#r28617443

**Comment by buchgr** [comment]
@nmittler yep he did https://github.com/netty/netty/pull/3649#issuecomment-94375705

**Comment by buchgr** [comment]
@nmittler yep he did https://github.com/netty/netty/pull/3649#issuecomment-94077741

**Comment by nmittler** [comment]
@buchgr ok great!

@Scottmitch I think you can pull the trigger on this.

**Comment by Scottmitch** [comment]
@normanmaurer @trustin - Good to cherry pick?

**Comment by normanmaurer** [comment]
@Scottmitch ship it :beers:

**Comment by Scottmitch** [comment]
Thanks for review all!

Cherry-picked 4.1( f812180) master (f1e122a)
